query/src/datasources/common/dal_metrics.rs

```diff
@@ -0,0 +1,137 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::SeekFrom;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::Poll;
+
+use common_dal::AsyncSeekableReader;
+use common_dal::DataAccessor;
+use common_dal::InputStream;
+use common_metrics::label_counter_with_val;
+use futures::Stream;
+
+pub const METRIC_DAL_READ_BYTES: &str = "dal.read.bytes";
+pub const METRIC_DAL_WRITE_BYTES: &str = "dal.write.bytes";
+
+struct InputStreamWrapper {
+    tenant_id: String,
+    cluster_id: String,
+    inner: InputStream,
+}
+
+impl InputStreamWrapper {
+    pub fn new(
+        tenant_id: impl Into<String>,
+        cluster_id: impl Into<String>,
+        input_stream: InputStream,
+    ) -> Self {
+        Self {
+            tenant_id: tenant_id.into(),
+            cluster_id: cluster_id.into(),
+            inner: input_stream,
+        }
+    }
+}
+
+pub struct DalWrapper {
+    tenant_id: String,
+    cluster_id: String,
+    inner: Arc<dyn DataAccessor>,
+}
+
+impl DalWrapper {
+    pub fn new(
+        tenant_id: impl Into<String>,
+        cluster_id: impl Into<String>,
+        inner: Arc<dyn DataAccessor>,
+    ) -> Self {
+        Self {
+            tenant_id: tenant_id.into(),
+            cluster_id: cluster_id.into(),
+            inner,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl DataAccessor for DalWrapper {
+    fn get_input_stream(
+        &self,
+        path: &str,
+        stream_len: Option<u64>,
+    ) -> common_exception::Result<InputStream> {
+        self.inner
+            .get_input_stream(path, stream_len)
+            .map(|input_stream| {
+                let r = InputStreamWrapper::new(&self.tenant_id, &self.cluster_id, input_stream);
+                Box::new(r) as Box<dyn AsyncSeekableReader + Unpin + Send>
+            })
+    }
+
+    async fn put(&self, _path: &str, _content: Vec<u8>) -> common_exception::Result<()> {
+        todo!()
+    }
+
+    async fn put_stream(
+        &self,
+        path: &str,
+        input_stream: Box<
+            dyn Stream<Item = Result<bytes::Bytes, std::io::Error>> + Send + Unpin + 'static,
+        >,
+        stream_len: usize,
+    ) -> common_exception::Result<()> {
+        self.inner
+            .put_stream(path, input_stream, stream_len)
+            .await
+            .map(|_| {
+                label_counter_with_val(
+                    super::dal_metrics::METRIC_DAL_WRITE_BYTES,
+                    stream_len as u64,
+                    self.tenant_id.as_str(),
+                    self.cluster_id.as_str(),
+                )
+            })
+    }
+}
+
+impl futures::AsyncRead for InputStreamWrapper {
```

Comment:
Looks like the InputStreamWrapper is a common stream who with metrics.
Would be nice move it to the common/streams as stream_metrics.rs?

Reply:
O, the metrics is about DAL.
I think we can make a new file in dal crate called `stream_metrics.rs` out from the dal_metrics file?

Reply:
:ok:

Reply:
> O, the metrics(METRIC_DAL_READ_BYTES...) is about DAL. I think we can make a new file in dal crate called `stream_metrics.rs` out from the dal_metrics file?

yeah, I am trying to move both DalWithMetrics and InputStreamWithMetrics into common/dal